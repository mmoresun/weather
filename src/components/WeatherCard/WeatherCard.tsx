import { useEffect, useState } from 'react';
import { IBigWeatherDataObj, ICityStateEntry } from '../../types/types';
import Card from '@mui/material/Card';
import CardActions from '@mui/material/CardActions';
import CardContent from '@mui/material/CardContent';
import Typography from '@mui/material/Typography';
import Box from '@mui/material/Box';
import ModalWindow from '../ModalWindow/ModalWindow';
import Button from '@mui/material/Button';

import { useDispatch } from 'react-redux';
// import { REMOVE_CITY } from '../../store/redusers/actions';
import { removeCityAction } from '../../store/redusers/cityReducer';

const WeatherCard = ({ ...props }) => {

    const dispatch = useDispatch();

    // deleting city from the list
    const deleteCity = (item: ICityStateEntry) => {

        // dispatch(
        //     {
        //         type: REMOVE_CITY,
        //         payload: item
        //     }
        // );

        const cityNeedToRemove = item;        

        dispatch(removeCityAction(cityNeedToRemove));
    }

    // manage the ModalWindow
    const [open, setOpen] = useState(false);
    const handleOpen = () => setOpen(true);
    const handleClose = () => setOpen(false);

    // trigger useEffect every time when the 'Refresh button is clicked'
    const [isRefresh, setIsreflesh] = useState(true);
    const handleRefresh = () => {
        setIsreflesh(!isRefresh)
    }

    const [cardData, setCardData] = useState({} as IBigWeatherDataObj);
    useEffect(() => {

        fetch(props.url)
            .then((response) => response.json())
            .then((data) => setCardData({
                name: data.name,
                country: data.sys.country,
                wind: data.wind.speed,
                clouds: data.weather[0].description,
                temp: Math.round(data.main.temp),
                feels_like: Math.round(data.main.feels_like),
                pressure: data.main.pressure,
                humidity: data.main.humidity,
                temp_min: data.main.temp_min,
                temp_max: data.main.temp_max,
                visibility: data.visibility,
                icon: data.weather[0].icon,
                wind_deg: data.wind.deg
            }));
        // .then((data) => console.log(data));
    }, [isRefresh, props.url])

    return (
        <>
            <Card sx={{ minWidth: 250, maxWidth: 300, margin: 1, boxShadow: 6, padding: '0 0 10px 0' }} variant="outlined">

                <CardContent>
                    <Typography variant="body2">City/Region: <b>{cardData.name}</b></Typography>
                    <Typography variant="body2">Temperature: <b>{cardData.temp}Â°C</b></Typography>
                    <Typography variant="body2">Wind: <b>{cardData.wind} m/s</b></Typography>
                    <Typography variant="body2">Sky: <b>{cardData.clouds}</b> </Typography>
                    <img src={`https://openweathermap.org/img/wn/${cardData.icon}@2x.png`} alt="" />
                </CardContent>

                <ModalWindow
                    open={open}
                    setOpen={setOpen}
                    handleOpen={handleOpen}
                    handleClose={handleClose}
                    cardData={cardData}
                />
                <CardActions sx={{ justifyContent: 'center' }}>
                    <Box sx={{ width: 220, display: 'flex', justifyContent: 'space-around', gap: .2 }}>
                        <Button
                            size="medium"
                            onClick={handleRefresh}
                        >Refresh &#x21bb;
                        </Button>
                        <Button
                            color="error"
                            size="medium"
                            onClick={() => deleteCity(props as ICityStateEntry)}
                        >Delete
                        </Button>
                    </Box>
                </CardActions>
            </Card>

        </>
    );
};

export default WeatherCard;