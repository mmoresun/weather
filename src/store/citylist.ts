
// initial state

import { v4 as uuid } from 'uuid';
import { ICityStateEntry } from '../types/types';

export const citylist: ICityStateEntry[] = [
  {
    url: 'https://api.openweathermap.org/data/2.5/weather?q=Kyiv&lang=en&appid=08f2a575dda978b9c539199e54df03b0&units=metric',
    id: uuid()
  },
  {
    url: 'https://api.openweathermap.org/data/2.5/weather?q=Warsaw&lang=en&appid=08f2a575dda978b9c539199e54df03b0&units=metric',
    id: uuid()
  },
  {
    url: 'https://api.openweathermap.org/data/2.5/weather?q=Prague&lang=en&appid=08f2a575dda978b9c539199e54df03b0&units=metric',
    id: uuid()
  },
  {
    url: 'https://api.openweathermap.org/data/2.5/weather?q=Antalya&lang=en&appid=08f2a575dda978b9c539199e54df03b0&units=metric',
    id: uuid()
  },
  {
    url: 'https://api.openweathermap.org/data/2.5/weather?q=Paris&lang=en&appid=08f2a575dda978b9c539199e54df03b0&units=metric',
    id: uuid()
  }
]
