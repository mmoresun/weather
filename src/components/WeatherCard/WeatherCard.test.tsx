
import renderer from 'react-test-renderer';
import WeatherCard from "./WeatherCard";
import { Provider } from 'react-redux';
import { store } from '../../store/store';

it('renders WeatherCard correctly', () => {
    const tree = renderer
      .create(<Provider store={store}><WeatherCard /></Provider>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });