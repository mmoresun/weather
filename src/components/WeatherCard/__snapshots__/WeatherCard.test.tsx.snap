// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders WeatherCard correctly 1`] = `
<div
  className="MuiPaper-root MuiPaper-outlined MuiPaper-rounded MuiCard-root css-193eagd-MuiPaper-root-MuiCard-root"
>
  <div
    className="MuiCardContent-root css-46bh2p-MuiCardContent-root"
  >
    <p
      className="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
    >
      Weather at 
      <b />
      :
    </p>
    <p
      className="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
    >
      ðŸŒ¡ Temperature: 
      <b>
        Â°C
      </b>
    </p>
    <p
      className="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
    >
      â˜´ Wind: 
      <b>
         m/s
      </b>
    </p>
    <p
      className="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
    >
      Sky: 
      <b />
       
    </p>
    <img
      alt=""
      src="https://openweathermap.org/img/wn/undefined@2x.png"
    />
  </div>
  <div>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall css-11qr2p8-MuiButtonBase-root-MuiButton-root"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      Details
    </button>
  </div>
  <div
    className="MuiCardActions-root MuiCardActions-spacing css-10r5to9-MuiCardActions-root"
  >
    <div
      className="MuiBox-root css-1py8c8m"
    >
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        Refresh â†»
      </button>
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textError MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textError MuiButton-sizeMedium MuiButton-textSizeMedium css-10fi992-MuiButtonBase-root-MuiButton-root"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        Delete
      </button>
    </div>
  </div>
</div>
`;
